### PR TITLE
[plugin.audio.ampache@matrix] 1.2.1+matrix.1

### PR DESCRIPTION
**Pull request by lusum**
### Description

**Add-on details:**

*General*

- Add-on name: Kodi Ampache plugin
-  Add-on ID: plugin.audio.ampache
- Version number: 1.2.1-matrix.1
- Kodi/repository version: matrix

Code location
- URL: https://github.com/ampache/ampache-xbmc-plugin

New plugin, official ampache server/nexcloud music plugin for kodi

1.2.1 (27.01.2021)

    Apply changes to be included in the official kodi repo plugin

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.
- If you see no activity on your PR after a week (so at least one weekend has passed) then please go to the #kodi-dev freenode IRC channel to reach out to the team

### CODE REVIEW

**Comment by enen92** [comment]
Are you somehow using our GitHub action scripts to do auto submissions? If so, you don't need to close and create a new pull request. Simply deleting a tag (locally and remote) and tagging it again with the same tag will push a new update to all your open pull requests

File: plugin.audio.ampache/resources/lib/ampache_connect.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                gcontext = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+                response = urllib.request.urlopen(req, context=gcontext, timeout=400)
+                xbmc.log("AmpachePlugin::handle_request: ssl",xbmc.LOGDEBUG)
+            else:
+                response = urllib.request.urlopen(req, timeout=400)
+                xbmc.log("AmpachePlugin::handle_request: nossl",xbmc.LOGDEBUG)
+        except urllib.error.HTTPError as e:
+            xbmc.log("AmpachePlugin::handle_request: HTTPError " +\
+                    repr(e),xbmc.LOGDEBUG)
+            xbmc.executebuiltin("ConnectionError" )
```
Comment (enen92) [inline plugin.audio.ampache/resources/lib/ampache_connect.py:89]: This builtin doesn't exist afaik..

File: plugin.audio.ampache/resources/lib/ampache_connect.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            raise self.ConnectionError
+        except urllib.error.URLError as e:
+            xbmc.log("AmpachePlugin::handle_request: URLError " +\
+                    repr(e),xbmc.LOGDEBUG)
+            xbmc.executebuiltin("ConnectionError" )
```
Comment (enen92) [inline plugin.audio.ampache/resources/lib/ampache_connect.py:94]: Same

File: plugin.audio.ampache/resources/lib/ampache_connect.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+            raise self.ConnectionError
+        except Exception as e:
+            xbmc.log("AmpachePlugin::handle_request: Generic Error "  +\
+                    repr(e),xbmc.LOGDEBUG)
+            xbmc.executebuiltin("ConnectionError" )
```
Comment (enen92) [inline plugin.audio.ampache/resources/lib/ampache_connect.py:99]: Same

File: plugin.audio.ampache/resources/lib/ampache_plugin.py [hunk 4]
```diff
@@ -0,0 +1,1053 @@
...
+                album_elem = node.find("album")
+                object_id = int(album_elem.attrib["id"])
+            except:
+                object_id = None
+        else:
+            object_id = int(node.attrib["id"])
+        image_url = check_get_art_url(node)
+        x = threading.Thread(target=art.get_art,args=(object_id,art_type,image_url,))
+        x.start()
+    x.join()
```
Comment (enen92) [inline plugin.audio.ampache/resources/lib/ampache_plugin.py:146]: I don't think this is right, X is a new thread instance each time you iterate on the for loop. You're only waiting for the last one to finish and not all of them. Save a thread reference into a list and execute join on all of them after the loop ends

**Review by enen92 — CHANGES_REQUESTED** [review]

**Comment by lusum** [comment]
>Are you somehow using our GitHub action scripts to do auto submissions? If so, you don't need to close and create a new pull request. Simply deleting a tag (locally and remote) and tagging it again with the same tag will push a new update to all your open pull requests

This is my first pull, so i fear to make a mess and i delete everything, but i can try

EDIT : i don't use action scripts, so i have to close and create a new pull request

>xbmc.executebuiltin("ConnectionError" )

Ok, i will cancel and i will test the new code

EDIT: done, i need some test before submit a new request. Also found a bug in notifications.

>I don't think this is right, X is a new thread instance each time you iterate on the for loop. You're only waiting for the last one to finish and not all of them. Save a thread reference into a list and execute join on all of them after the loop ends


Ok, thanks, i will make the change, it doesn't change the overall behaviour, but it is more correct.

EDIT: correct also this one, also found a bug that made this function never called

If they aren't any new changes to do, i will test some days and i will close this pull and i will made a new one

**Comment by lusum** [comment]
close now,

**Pull request closed without merging**